### PR TITLE
app: Make OperatioalDeviceProxy owned by each user

### PR DESCRIPTION
**Pull request by gjc13**
#### Problem

To avoid accidentaly closing other user's Operational Device, we make
OperatioalDeviceProxy owned by each user. The OperationalDeviceManager
will now try to share the SecureSession when possible.

#### Change overview

* Rename `CASESessionManager` to `OperationalDeviceManager`
* Create a new `OperationalDeviceProxy` for each `AcquireDevice` call.
* Share the underlying secure session when possible.

#### Testing

Locally tested by pairing and a patch on the binding branch.
Pairing & control tested by the CI.

### CODE REVIEW

File: src/app/OperationalDeviceManager.cpp [hunk 1]
```diff
@@ -0,0 +1,181 @@
+/*
```
Comment (bzbarsky-apple) [inline src/app/OperationalDeviceManager.cpp:1]: Please do the rename in a separate PR from substantive changes that are big enough that git does not detect the rename.  Otherwise reviewing is a huge pain....

@pan-apple Please look this over and see whether it fits in with the planned evolution of this class and CASE sessions?

File: src/app/OperationalDeviceManager.h [hunk 2]
```diff
@@ -71,35 +72,38 @@ class CASESessionManager : public Dnssd::ResolverDelegate
...
+     *
      * If the connection is already established, the `onConnection` callback will be immediately called.
      */
-    CHIP_ERROR FindOrEstablishSession(PeerId peerId, Callback::Callback<OnDeviceConnected> * onConnection,
-                                      Callback::Callback<OnDeviceConnectionFailure> * onFailure);
+    CHIP_ERROR AcquireDevice(PeerId peerId, Callback::Callback<OnDeviceConnected> * onConnection,
+                             Callback::Callback<OnDeviceConnectionFailure> * onFailure);
 
-    OperationalDeviceProxy * FindExistingSession(PeerId peerId);
+    void ReleaseDevice(OperationalDeviceProxy * device);
```
Comment (mrjerryjohns) [inline src/app/OperationalDeviceManager.h:89]: If we're going to have a manager-like object, I'd very much prefer if we limited interactions with it to just during initialization/allocation of the `OperationalDeviceProxy`. There-after, clients should just be engaging with the latter, not the former.

File: src/app/OperationalDeviceManager.cpp [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+namespace chip {
+
+CHIP_ERROR OperationalDeviceManager::AcquireDevice(PeerId peerId, Callback::Callback<OnDeviceConnected> * onConnection,
+                                                   Callback::Callback<OnDeviceConnectionFailure> * onFailure)
+{
+    Dnssd::ResolvedNodeData resolutionData;
+
+    bool nodeIDWasResolved = (mConfig.dnsCache != nullptr && mConfig.dnsCache->Lookup(peerId, resolutionData) == CHIP_NO_ERROR);
+
+    OperationalDeviceProxy * device = FindActiveDevice(peerId);
```
Comment (mrjerryjohns) [inline src/app/OperationalDeviceManager.cpp:41]: What happens if we are in the process of establishing a CASE session with another `OperationalDeviceProxy` when  we get here? Wouldn't that result in two separate, but similar CASE sessions (similar = to the same target) being present?

File: src/app/OperationalDeviceManager.cpp [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    return Loop::Continue;
+}
+
+void OperationalDeviceManager::HandleDeviceConnected(void * context, OperationalDeviceProxy * device)
+{
+    OperationalDeviceManager * manager = static_cast<OperationalDeviceManager *>(context);
+    if (!manager->mNotifyingConnectionSucess)
+    {
+        manager->mNotifyingConnectionSucess = true;
+        manager->mConfig.devicePool->ForEachDevice(device, NotifyDeviceConnected);
```
Comment (mrjerryjohns) [inline src/app/OperationalDeviceManager.cpp:104]: Why is this notifying every single active object? Shouldn't it just be notifying device's that 'match'?

File: src/app/OperationalDeviceManager.h [hunk 5]
```diff
@@ -71,35 +72,38 @@ class CASESessionManager : public Dnssd::ResolverDelegate
...
 
-    virtual ~CASESessionManager() { mDNSResolver.Shutdown(); }
+    virtual ~OperationalDeviceManager() { mDNSResolver.Shutdown(); }
 
     /**
      * Find an existing session for the given node ID, or trigger a new session request.
      * The caller can optionally provide `onConnection` and `onFailure` callback objects. If provided,
      * these will be used to inform the caller about successful or failed connection establishment.
+     *
+     * The user shall call ReleaseDevice to free the device in the failure callback.
```
Comment (mrjerryjohns) [inline src/app/OperationalDeviceManager.h:82]: Who 'owns' the `OperationalDeviceProxy` object returned in the callback? Who's responsible for free'ing it during success? During failure?
